stopes/pipelines/asr_bleu/transcribe_audio.py

```diff
@@ -0,0 +1,213 @@
+    
+    def requirements(self) -> Requirements:
+        return Requirements(
+            nodes=1,
+            tasks_per_node=1,
+            gpus_per_node=0,
+            cpus_per_task=1,
+            timeout_min=720,
+        )
+                  
+    @torch.inference_mode()
+    def _load_audiofile(self, audio_path: str, sample_rate: int, normalize_input: bool) -> torch.Tensor:
+        """
+        Load the audio files and apply resampling and normalization
+
+        Args:
+            audio_path: the audio file path
+            sample_rate: the target sampling rate for the ASR model
+            normalize_input: the bool value for the ASR model
+
+        Returns:
+            audio_waveform: the audio waveform as a torch.Tensor object
+        """
+        audio_waveform, sampling_rate = torchaudio.load(audio_path)
+        if audio_waveform.dim == 2:
+            audio_waveform = audio_waveform.mean(-1)
+        if sample_rate != sampling_rate:
+            audio_waveform = torchaudio.functional.resample(audio_waveform, sampling_rate, sample_rate)
+        if normalize_input:
+            # following fairseq raw audio dataset
+            audio_waveform = torch.nn.functional.layer_norm(audio_waveform, audio_waveform.shape)
+
+        return audio_waveform
+
+    @torch.inference_mode()
+    def _compute_emissions(self, audio_input: torch.Tensor, asr_model: fairseq.models.FairseqEncoder, cuda) -> torch.Tensor:
+        """
+        Compute the emissions for either fairseq or huggingface asr model
+
+        Args:
+            audio_input: the input audio waveform
+            asr_model: the ASR model to use for computing emissions
+
+        Returns:
+            emissions: the logits of the encoded prediction.
+        """
+
+        if cuda:
+            audio_input = audio_input.to("cuda")
+        if isinstance(asr_model, fairseq.models.wav2vec.wav2vec2_asr.Wav2VecCtc):
+            padding_mask = lengths_to_padding_mask(torch.tensor([audio_input.numel()]))
+            emissions = asr_model.w2v_encoder(audio_input, padding_mask)["encoder_out"].transpose(0, 1)
+        else:
+            emissions = asr_model(audio_input).logits
+
+        return emissions
+
+
+    def _decode_emissions(self,emissions: torch.Tensor, decoder, post_process_fn) -> str:
+        """
+        Decode the emissions and apply post-process functions
+
+        Args:
+            emissions: the input Tensor object
+            decoder: ASR decoder
+            post_process_fn: post-process function for the ASR output
+
+        Returns:
+            hypo: the str as the decoded transcriptions
+        """
+
+        emissions = emissions.cpu()
+        results = decoder(emissions)
+
+        # assuming the lexicon-free decoder and working with tokens
+        hypo = decoder.idxs_to_tokens(results[0][0].tokens)
+        hypo = post_process_fn(hypo)
+
+        return hypo
+
+    def _merge_tailo_init_final(self, text):
```

Comment:
This method doesn't seem to reuse any properties or methods of the `TranscribeAudio`, so probably it should be turned into a standalone function. 

Also, it's totally unclear what it actually does. Maybe, a few examples of inputs and outputs could illustrate it?

Reply:
This function is for post processing the ASR output to merge the phoneme initial and finals back to syllable units and it is specifically for Hokkien, details in https://github.com/fairinternal/ust_common/pull/47